20181114_Connecting_R_to_SQL/Connecting_R_to_SQL.Rmd

```diff
@@ -0,0 +1,462 @@
+most of the data we have is stored in SQL).
+
+We also need the `odbc` package as that provides the driver for SQL databases.
+
+
+### Advantages of `DBI` and `odbc` over `RODBC`
+
+I had previously always used the `RODBC` package to connect to databases.  Reasons for the move to using `DBI` and `odbc` include:
+
+1. Enables use of the RStudio Connections Pane (see below for more info)
+
+2. I found the `DBI`/`odbc` solution much quicker for writing out to SQL
+
+3. You can use `dbplyr` when your data is too big to work with in memomry with
+`dplyr`, or if you are not familiar with SQL.
+
+
+### Create a connection
+
+Start by creating a connection to the database:
+
+```{r}
+
+conn <- DBI::dbConnect(
+  odbc::odbc()
+  , .connection_string = "driver={SQl Server};
+                          server=TS02; 
+                          database=RCS_TestArea"
+  )
+
+```
+
+
+When you have done this you can see the details of the connection to the database in the Connections pane where you usually see the Environment:
+<br>
+<br>
+![](images/r_conns.png)
+
+You can open this up to see more detail on the tables within the database, under **dbo** we just have the one table **pcd_sample**:
+<br>
+<br>
+![](images/r_conns_tables.png)
+
+### Read in data
+
+If you want to send a **SELECT** query to the database you can use the **dbGetQuery()** function. 
+
+For example, this query will bring in just the records where the postcode starts AL1:
+
+```{r}
+
+pcd_al1 <- DBI::dbGetQuery(
+  conn
+  , "SELECT * FROM [dbo].[pcd_sample] WHERE PCD8 like 'AL1%'"
+  )
+
+nrow(pcd_al1)
+
+```
+
+Or you could count the number of different Local Authority Districts:
+
+```{r}
+
+DBI::dbGetQuery(
+  conn
+  , "SELECT COUNT (distinct LAD11NM) FROM [dbo].[pcd_sample]"
+  )
+
+```
+
+If you have a more complicated query you can use `str_c` from the `stringr`
+package to concatenate the string:
+
+
+```{r}
+
+sql_table <- "[dbo].[pcd_sample]"
+
+sql_query <- stringr::str_c(
+  "SELECT COUNT(distinct LAD11NM) FROM"
+  , sql_table
+  )
+
+DBI::dbGetQuery(
+  conn
+  , sql_query
+  )
+
+```
+
+### Joining data
+
+If this example I have a small dataset containing postcodes and I want to add 
+local authority information to it by joining it with the **pcd_sample** data.
+
+```{r}
+
+pokemon <- readr::read_csv(
+  file = "data/pokemon_pcd.csv"
+  )
+
+knitr::kable(pokemon[1:6, ])
+
+```
+
+#### Option 1 - Read the whole SQL table into R
+
+I could read all the Postcode data in to R ready to do the joining:
+
+```{r}
+start <- Sys.time()
+
+pcd <- DBI::dbGetQuery(
+  conn
+  , "SELECT * FROM [dbo].[pcd_sample]"
+  )
+
+```
+
+
+This takes:
+
+```{r}
+Sys.time() - start
+```
+
+It is potentially slow due to the size of the data. However, I have trimmed this data set down to `r format(nrow(pcd), big.mark = ",")` rows, a mere 1% of the full postcode lookup available.
+
+
+#### Option 2 - Read just the rows of the SQL table you need into R

```

Comment:
nah - CA didn't have one, dunno why. Doesn't really help to compare since when I rendered it the bit pulling in all the data is actually faster than option 3